### PR TITLE
Allow formatting link parameter values as ISO date #2487

### LINKED ISSUE #2487: [Feature]: Allow formatting template variables in link patterns
### Requirement

As a user of jaeger and another logging system (in my case OpenSearch + OpenSearch dasbhoards) I would like to link from traces in my Jaeger UI to logs in my OpenSearch Dashboards UI while filtering on a specific time range.

### Problem

I can currently add a filter on trace ID easily (and this is wonderful, thank you everyone who worked on this 😄). I would also like to apply a time range filter into OpenSearch Dashboards to filter to the right time period for my trace, but given that the `startTime` from Jaeger is a time value in microseconds I am unable to do so.

### Proposal

In Grafana I can do this with a data link using the format `from:'${__from:date}'` where `:date` formats the value as a date.

My link pattern in Jaeger looks like
```json
{
  "type": "traces",
  "url": "https://mydashboards.com/_dashboards/app/discover#/?_g=(time:(from:'#{startTime}',to:'#{endTime}'))&_a=(index:filebeat-all,query:(language:kuery,query:'#{traceID}'))",
  "text": "Logs"
}
```

A similar syntax to allow formatting time values to dates (or if there is something else that is compatible with OpenSearch Dashboards that I don't know of that would satisfy my use case).

Let me know if I can provide any more context

### Open questions

_No response_

### PR DESCRIPTION
**Pull request by drewcorlin1**
## Which problem is this PR solving?
Resolves #2487

## Description of the changes
Add the ability to format link variables.

## How was this change tested?
Unit tests and by running it locally with the following UI config
```javascript
const DEFAULT_CONFIG = {
  "linkPatterns": [
    {
      "type": "traces",
      "url": "https://myOpenSearch.com/_dashboards/app/discover#/?_g=(time:(from:'#{startTime | epoch_micros_to_date_iso}',to:'#{endTime | epoch_micros_to_date_iso}'))&_a=(index:filebeat-all,query:(language:kuery,query:'#{traceID}'))",
      "text": "Logs"
    }
  ]
};
```

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

File: packages/jaeger-ui/src/utils/link-formatting.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      return microsSinceEpoch;
+    }
+
+    return new Date(microsSinceEpoch / 1000).toISOString();
+  },
+};
+
+export function getParameterAndFormatter(parameter: string) {
+  const [parameterName, formatFunctionName] = parameter.split('|').map(part => part.trim());
+  const formatFunction = formatFunctions[formatFunctionName];
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/link-formatting.tsx:32]:
```suggestion
  if (!formatFunctionName) {
    return { parameterName, null };
  }
  const formatFunction = formatFunctions[formatFunctionName];
```

File: packages/jaeger-ui/src/utils/stringSupplant.tsx [hunk 2]
```diff
@@ -20,7 +22,8 @@ export function encodedStringSupplant(
   map: Record<string, string | number | undefined>
 ) {
   return str.replace(PARAMETER_REG_EXP, (_, name) => {
-    const mapValue = map[name];
+    const { parameterName } = getParameterAndFormatter(name);
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/stringSupplant.tsx:25]: why is it ok to ignore formatter here?

File: packages/jaeger-ui/src/utils/link-formatting.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { Trace } from '../types/trace';
+
+const formatFunctions: Record<string, (value: Trace[keyof Trace]) => string | Trace[keyof Trace]> = {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/link-formatting.tsx:17]: can we add unit tests in this package?

File: packages/jaeger-ui/src/utils/stringSupplant.tsx (hunk 2 above)
Comment (drewcorlin1) [reply packages/jaeger-ui/src/utils/stringSupplant.tsx]: Added

File: packages/jaeger-ui/src/utils/link-formatting.tsx (hunk 1 above)
Comment (drewcorlin1) [reply packages/jaeger-ui/src/utils/link-formatting.tsx:39]: Updated

File: packages/jaeger-ui/src/utils/link-formatting.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2020 The Jaeger Authors.
```
[the remark this answers is not in the document]
Comment (drewcorlin1) [reply packages/jaeger-ui/src/utils/link-formatting.tsx:1]: Updated

File: packages/jaeger-ui/src/utils/link-formatting.tsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/link-formatting.tsx:1]:
```suggestion
// Copyright (c) 2024 The Jaeger Authors.
```

**Review by yurishkuro — APPROVED** [review]

**Comment by yurishkuro** [comment]
please add this to the docs:
  * repo jaegertracing/documentation
  * files
    * content/docs/next-release-v2/frontend-ui.md
    * content/docs/next-release/frontend-ui.md

**Comment by drewcorlin1** [comment]
Planning on updating those docs this week

**Comment by drewcorlin1** [comment]
Doc updates are here https://github.com/jaegertracing/documentation/pull/807

File: packages/jaeger-ui/src/utils/stringSupplant.tsx (hunk 2 above)
Comment (drewcorlin1) [reply packages/jaeger-ui/src/utils/stringSupplant.tsx]: Updated to respect the formatter

**Review by yurishkuro — APPROVED** [review]

**Pull request merged**